Improve Live Reduction Documentation (#48)
mantidproject/livereduce: docs/architecture.md

## System Components

The live reduction system has several interconnected components:

```mermaid
flowchart TD
    DAS["Data Acquisition System (DAS)<br/>Sends variable-sized data chunks"]
    Listener["Live Data Listener<br/>(SNSLiveEventDataListener, ISISHistoDataListener, KafkaListener)<br/>Receives chunks and buffers them in memory"]
    Mantid["Mantid StartLiveData / MonitorLiveData<br/>Orchestrates the processing and accumulation cycle"]
    ProcScript["Processing Script<br/>(proc.py)<br/>Runs on each chunk received"]
    PostScript["Post-Processing Script<br/>(post_proc.py)<br/>Runs on accumulated data at intervals"]
    Daemon["livereduce Daemon<br/>Manages the entire workflow, monitors for script changes,<br/>watches memory usage, handles restarts"]
    Watchdog["livereduce_watchdog<br/>Monitors the daemon's log file and restarts it if unresponsive"]

    DAS --> Listener
    Listener --> Mantid
    Mantid --> ProcScript
    Mantid --> PostScript
    ProcScript --> Daemon
    PostScript --> Daemon
    Daemon --> Watchdog
```

### Component Descriptions

**DAS (Data Acquisition System)**
- Collects neutron event or histogram data from detectors

> - Streams data to listeners

Reviewer: Do we want this to be a tiny bit more explicit

```suggestion
- Streams data to registered listeners
```

Author: Added

- Chunk sizes vary based on event rate and detector activity
- You cannot control chunk size - DAS sends what it has buffered

**Live Data Listener**
- Mantid components that connect to DAS and receive data chunks
- Different listeners support different protocols (TCP, Kafka) and formats (events, histograms)
- Examples: `SNSLiveEventDataListener`, `ISISHistoDataListener`, `KafkaEventListener`

**StartLiveData / MonitorLiveData**
- Mantid algorithms managing the live data workflow
- `StartLiveData`: Initializes connection and processing
- `MonitorLiveData`: Runs in loop, calling processing scripts periodically

**Processing Scripts**
- Python scripts defining data transformations
- `proc.py`: Processes each incoming chunk
- `post_proc.py`: Processes accumulated data
- Instrument-specific and user-written

**livereduce Daemon**
- Systemd service wrapping the entire workflow
- Provides automatic restart on errors
- Monitors memory usage
- Detects script changes via inotify file watching

> **livereduce_watchdog**

Reviewer: Please mention that this is a systemd service too

Author: Added.

- Independent monitoring service
- Checks if main daemon is writing to log file
- Restarts main service if unresponsive

## Data Flow

Understanding data flow is key to effective live reduction:

### 1. DAS Sends Chunks
- DAS transmits data in variable-sized chunks
- Timing and size determined by DAS, not controllable
- Chunk size varies based on event rate and buffering

### 2. Listener Buffers
- Listener accumulates events/histograms into buffer
- Passes to Mantid when threshold or timeout reached
- Buffer size configurable per listener type

### 3. Processing Executes
- If `proc.py` exists, runs on each chunk received
- Processed chunk updates `output` workspace
- Blocks until processing completes

### 4. Accumulation Occurs
Mantid accumulates chunks based on `AccumulationMethod`:

> - **Add**: Sum processed chunks together (default, good for event data)

Reviewer: Add runs the `Plus` algorithm on the current output from the proc and the accumulation workspace. For histograms, this changes values. For events, this appends the new events to the EventList and grows forever.

Author: Added

- **Replace**: Each chunk replaces previous data (good for histogram monitoring)

> - **Append**: Stack spectra from each chunk (workspace grows)

Reviewer: I've never used method, but according to the doxygen "Accumulate the data by appending the spectra into the the output workspace." This would grow the number of spectra in the workspace.

Author: Understood, updated to reflect this.

### 5. Post-Processing Executes

> - Every `UpdateEvery` seconds (default 30)

Reviewer: ```suggestion
- At minimum, `UpdateEvery` seconds (default 30). Will be slower if the "proc" script takes longer than this setting.
```

Author: Added

- If `post_proc.py` exists, runs on `accumulation` workspace

> - Non-blocking - doesn't prevent chunk processing

Reviewer: This is incorrect

Author: Fixed

### 6. Memory Monitoring
- If enabled, daemon checks memory usage periodically

> - Restarts processing if limits exceeded

Reviewer: ```suggestion
- Restarts the daemon if limits exceeded
```

Author: Added

> - Prevents system crashes from memory exhaustion

Reviewer: ```suggestion
- Terminates the daemon before [OOM Killer](https://learn.redhat.com/t5/Platform-Linux/Out-of-Memory-Killer/td-p/48828) does. This allows for logging the issue.
```

Author: Added

### 7. Output Generation
- Post-processing typically generates files, plots, web content
- Scientists view results in real-time
- Output location configurable per instrument

## Processing vs Post-Processing

> The two-stage model separates lightweight per-chunk operations from heavier accumulated data analysis:

Reviewer: What is "light" versus "heavy" is entirely dependent on what people write/provide for the proc/post-proc steps. I like that you are pointing out that people can balance the two steps with each other.

Author: Understood

### Processing Script (`reduce_<INSTRUMENT>_live_proc.py`)

> **Purpose**: Fast, lightweight operations on each incoming chunk

Reviewer: ```suggestion
**Purpose**: operations on each incoming chunk
```

Author: Keep expectations at bay haha

> **When it runs**: Every time a chunk arrives (unpredictable timing)

Reviewer: After the previous proc step or `UpdateEvery` seconds. Whichever is slower.

Author: Understood, corrected

**Available workspaces**:
- `input`: The incoming data chunk (workspace)

> - `output`: Where to put results (workspace name as string)

Reviewer: There is a gotcha here. You get the workspace pointer as the variable `output` which is also the workspace's name. This reference/pointer has to be overwritten by your script.

Author: Added

> **Performance considerations**:

Reviewer: Live reduction will keep accumulating data and will wait to rerun the step. Also, DAS has its own caching and keeps track of events that have been sent already. Does not need to be fast, but faster is better.

Author: Fixed

- Runs frequently, must be fast
- Blocks receiving new chunks
- Avoid file I/O

**Typical operations**:
- Basic rebinning
- Spectrum summing
- Simple filtering
- Unit conversions
- Detector grouping

**Example**:
```python
from mantid.simpleapi import Rebin, SumSpectra

# Sum all spectra in the chunk
SumSpectra(InputWorkspace=input, OutputWorkspace=output)
# Rebin to coarser bins for faster plotting
Rebin(InputWorkspace=output, OutputWorkspace=output,
      Params="300,20,17000")
```

### Post-Processing Script (`reduce_<INSTRUMENT>_live_post_proc.py`)

**Purpose**: Comprehensive analysis on accumulated data

> **When it runs**: Every `UpdateEvery` seconds (default 30s)

Reviewer: ```suggestion
**When it runs**: After every "proc" step is acculuated
```

Author: Fixed

**Available workspaces**:
- `accumulation`: All data accumulated so far

> - Any workspaces created during processing

Reviewer: There is also an `output` workspace, just like the "proc" step

Author: Understood, updated

**Performance considerations**:
- Runs less frequently
- Can be computationally intensive
- File I/O appropriate here

**Typical operations**:
- Background subtraction

> - Normalization

Reviewer: ```suggestion
- Normalization by measurement time or accumulated proton charge
```

Author: Fixed

- Peak fitting
- File writing (NeXus, ASCII, images)
- Web service updates
- Complex calculations

**Example**:
```python
from mantid.simpleapi import SaveAscii, SaveNexus

# Save accumulated data to shared directory
output_dir = "/SNS/INSTR/IPTS-12345/shared/livereduce"
SaveNexus(InputWorkspace="accumulation",
          Filename=f"{output_dir}/live_data.nxs")
SaveAscii(InputWorkspace="accumulation",
          Filename=f"{output_dir}/live_data.txt")
```

## Best Practices

### Script Performance
- Keep processing scripts lightweight
- Put file I/O in post-processing only

> - Test with realistic data rates before deploying

Reviewer: It would be good to add a comment here about proc scripts that output EventWorkspace can create very large accumulation workspaces. This choice is either to have the proc script return a histogram, or histogram early in the post-proc.

Author: Comment added

### Memory Management
- Consider event preservation carefully - very memory intensive
- Use `PreserveEvents=False` when possible
- Monitor memory usage in production
- Configure appropriate `system_mem_limit_perc`

### External Integrations

> - Use post-processing for database updates

Reviewer: Do you mean external databases?

Author: Yes, I do. Updated

- Put web service calls in post-processing
- Handle errors gracefully in post-processing scripts

### Testing
- Use fake data servers for development
- Test memory usage under load
- Verify scripts before production deployment

## Related Documentation

> 

Reviewer: Please add links to StartLiveData, MonitorLiveData, and LoadLiveData in the mantid user docs

Author: Links added.
